File: data-structures/queue/Queue.test.js
const Queue = require('./Queue.js')

const testNewQueue = () => {
    const queue = new Queue()

    console.assert(queue.isEmpty(), 'expect queue to be empty')
}

const testQueueEnqueue = () => {
    const queue = new Queue()
    queue.enqueue(1)

    console.assert(!queue.isEmpty(), 'expect queue to have some values')
}

const testQueuePeek = () => {
    const queue = new Queue(), expected = 1
    queue.enqueue(expected)
    const value = queue.peek()

    console.assert(!queue.isEmpty(), 'expect queue to have some values')
    console.assert(value === expected, `expect to be: ${expected}, got instead: ${value}`)
}

const testQueueDequeue = () => {
    const queue = new Queue(), expected = 1
    queue.enqueue(expected)
    const value = queue.dequeue()

    console.assert(queue.isEmpty(), 'expect queue to be empty')
    console.assert(value === expected, `expect to be: ${expected}, got instead: ${value}`)
}

testNewQueue()
testQueueEnqueue()
testQueuePeek()
testQueueDequeue()

File: data-structures/queue/queue_test.go
package queue

import (
	"testing"
)

func TestQueue_IsEmpty(t *testing.T) {
	queue := New()
	if !queue.IsEmpty() {
		t.Error("expect Queue to be empty")
	}
}

func TestQueue_Enqueue(t *testing.T) {
	queue := New()

	queue.Enqueue(1)
	if queue.IsEmpty() {
		t.Error("expect Queue to have some values")
	}
}

func TestQueue_Dequeue(t *testing.T) {
	queue := New()

	queue.Enqueue(2)
	queue.Enqueue(3)

	if expected, val := 2, queue.Dequeue(); expected != val {
		t.Errorf("expect enqueued value to be [%d], got instead [%d]", expected, val)
	}

	if expected, val := 3, queue.Dequeue(); expected != val {
		t.Errorf("expect enqueued value to be [%d], got instead [%d]", expected, val)
	}

	if !queue.IsEmpty() {
		t.Error("expect Queue to be empty")
	}
}

func TestQueue_Peek(t *testing.T) {
	queue := New()

	queue.Enqueue(2)
	queue.Enqueue(3)

	if expected, val := 2, queue.Peek(); expected != val {
		t.Errorf("expect enqueued value to be [%d], got instead [%d]", expected, val)
	}

	if expected, val := 2, queue.Peek(); expected != val {
		t.Errorf("expect enqueued value to be [%d], got instead [%d]", expected, val)
	}
}
